src/DataStreams/CsvStream.php

```diff
@@ -0,0 +1,302 @@
+<?php
+
+/*
+ *	Copyright 2015 RhubarbPHP
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+namespace Rhubarb\Crown\DataStreams;
+
+require_once __DIR__ . '/DataStream.php';
+
+use Rhubarb\Crown\Exceptions\EndOfStreamException;
+
+class CsvStream extends DataStream
+{
+    private $filePath;
+
+    private $fileStream = null;
+
+    private $writable = false;
+
+    private $headers = null;
+
+    private $hasHeaders = true;
+
+    private $needToWriteHeaders = true;
+
+    private $enclosure = "\"";
+
+    private $delimiter = ",";
+
+    private $remnantBuffer = "";
+
+    public function __construct($filePath)
+    {
+        $this->filePath = $filePath;
+
+        parent::__construct();
+    }
+
+    public function setHeaders($headers)
+    {
+        $this->headers = $headers;
+        $this->hasHeaders = true;
+    }
+
+    private function readHeaders()
+    {
+        $this->close();
+
+        if (!file_exists($this->filePath)) {
+            $this->needToWriteHeaders = true;
+            $this->headers = [];
+
+            return;
+        }
+
+        $this->fileStream = fopen($this->filePath, "r");
+
+        $rawCsvData = $this->readCsvLine();
+
+        $this->headers = $rawCsvData;
+        $this->needToWriteHeaders = false;
+    }
+
+    private function readCsvLine()
+    {
+        if (feof($this->fileStream) && ($this->remnantBuffer == "")) {
+            throw new EndOfStreamException();
+        }
+
+        $inEnclosure = false;
+        $readFullLine = false;
+
+        $values = [];
+
+        $addValue = function (&$value) use (&$values) {
+            $values[] = utf8_encode(str_replace("@@@escapedenclosure@@@", $this->enclosure, $value));
+            $value = "";
+        };
+
+        $valueBuffer = "";
+
+        while (!$readFullLine) {
+            if (feof($this->fileStream) && ($this->remnantBuffer == "")) {
+                break;
+            }
+
+            // Read the next set of bytes and prepend with any remaining buffer from the last read.
+
+            if (strlen($this->remnantBuffer) < 1024) {
+                $csvData = $this->remnantBuffer . fread($this->fileStream, 1024);
+            } else {
+                $csvData = $this->remnantBuffer;
+            }
+
+            if ($this->enclosure != "") {
+                $csvData = str_replace($this->enclosure . $this->enclosure, "@@@escapedenclosure@@@", $csvData);
```

Comment:
It might be better to add knowledge of an escape character to the parser for cvs's that don't comply with rfc4180 and to avoid replacing a string which we assume won't be in the file (we can't safely make this assumption if we're making this publicly available) - could instead test for escape character followed by enclosure when inside an enclosure.